File: Solutions/DailyCodingProblems/DCP-1288/Solution.py
def balance(s: str) -> str:
    res, open_count = [], 0
    for char in s:
        if char == '(':
            open_count += 1
            res.append(char)
        elif char == ')':
            if open_count > 0:
                open_count -= 1
            else:
                res.append('(')
            res.append(char)
    return "".join(res) + ")" * open_count

if __name__ == "__main__":
    print(balance("(()"))     # Output: (())
    print(balance("))()("))   # Output: ()()()()

File: Solutions/DailyCodingProblems/DCP-1288/Solution.cs
using System;
using System.Text;

class Program
{
    public static string Balance(string s)
    {
        var res = new StringBuilder();
        int openCount = 0;

        foreach (char ch in s)
        {
            if (ch == '(')
            {
                openCount++;
                res.Append(ch);
            }
            else if (ch == ')')
            {
                if (openCount > 0)
                {
                    openCount--;
                }
                else
                {
                    res.Append('(');
                }
                res.Append(ch);
            }
            else
            {
                res.Append(ch);
            }
        }

        // Append remaining closing parentheses for any unmatched opening ones
        res.Append(')', openCount);

        return res.ToString();
    }

    static void Main()
    {
        Console.WriteLine(Balance("(()"));     // Output: (())
        Console.WriteLine(Balance("))()("));   // Output: ()()()()
    }
}